spot-contracts/contracts/PerpetualTranche.sol

```diff
@@ -978,20 +995,53 @@ contract PerpetualTranche is
         return _reserves.contains(address(token));
     }
 
+    /// @dev Queries the current subscription state of the perp and vault systems.
+    function _querySubscriptionState() private view returns (IFeePolicy.SubscriptionParams memory s) {
+        s.perpTVL = _reserveValue();
+        s.vaultTVL = IVault(vault).getTVL();
+        (s.perpTR, s.vaultTR) = _depositBond.getSeniorJuniorRatio();
+        return s;
+    }
+
     /// @dev Calculates the total value of all the tranches in the reserve.
-    ///      Value of each reserve tranche is calculated as = trancheBalance . tranchePrice.
+    ///      Value of each reserve tranche is denominated in the underlying collateral.
     function _reserveValue() private view returns (uint256) {
-        uint256 totalVal = _reserveAt(0).balanceOf(address(this));
+        IERC20Upgradeable underlying_ = _reserveAt(0);
+        uint256 totalVal = underlying_.balanceOf(address(this));
         for (uint256 i = 1; i < _reserveCount(); i++) {
-            IERC20Upgradeable token = _reserveAt(i);
-            totalVal += token.balanceOf(address(this)).mulDiv(_tranchePrice(ITranche(address(token))), UNIT_PRICE);
+            ITranche tranche = ITranche(address(_reserveAt(i)));
+            totalVal += _computeReserveTrancheValue(
+                tranche,
+                IBondController(tranche.bond()),
+                underlying_,
+                tranche.balanceOf(address(this)),
+                true
+            );
         }
         return totalVal;
     }
 
-    /// @dev Fetches price of a given tranche from the pricing strategy.
-    function _tranchePrice(ITranche t) private view returns (uint256) {
-        return pricingStrategy.computeTranchePrice(t);
+    /// @dev Computes the value of the given amount reserve tranche tokens (i.e ones already accepted in the reserve or to be accepted),
+    ///      based on it's current CDR.
+    ///      NOTE: Callers should round up when valuing reserve assets and round down for incoming assets.
```

Comment:
I think of rounding up value of outgoing tokens and rounding down value of incoming tokens.

It's less clear to me, without knowing the caller's use case, whether we should always round up the value of reserve assets, though. Do you have a simple rationale?

Reply:
When tokens enter/exit the system, we compare the value of the incoming token (for which we round down) wrt the value of assets in the reserve (for which we round up).

In perp, tokens enter/exit the system based on value only on mint and rollover.
Mint -> valueOf(trancheIn, ROUND_DOWN), valueOf(reserve, ROUND_UP)
Rollover -> valueOf(trancheIn, ROUND_DOWN), valueOf(reserveToken, ROUND_UP)

Reply:
So basically, the only time we care about the /value/ of reserve assets is when we're trading them out for different incoming assets... I guess that makes sense. Otherwise, internally it only needs to deal with token values.